---
toc-depth: 2
---

![](img/banner-min.png)

<div style="text-align: justify">

On this website you can find documentation about software commonly used in bioinformatic data analyses as well as tutorials about various bioinformatic subjects. On this webpage you can find software organized by topic and for each topic you find a list of commonly used software tools.

If you are working at the University of Amsterdam (UvA) Institute for Biodiversity and Ecosystem Dynamics (IBED) and want to know more about what computational resources are available, please also visit the [computational support teams website](https://ibed.uva.nl/facilities/computational-facilities/ibed-computational-support-team/ibed-computational-support-team.html) and our website with more [computation resources](https://computational-resources-uva-ibed-cs-general-20079b9f20f96949e05.gitlab.io/)

Please, be aware that this page is a work in progress and will be slowly updated over time. If you want to add additional information or feel that something is missing feel free to send an email to [n.dombrowski\@uva.nl](mailto:n.dombrowski@uva.nl){.email}.

## Useful tutorials

The Carpentries teaches workshops around the world on the foundational skills to work effectively and reproducibly with data and code and they are an excellent resource to check out if you want to get started with bioinformatics

- [The software carpentries](https://software-carpentry.org/lessons/) provides tutorials on:
  - Bash
  - Git 
  - Python
  - R
- [Data carpentries](https://datacarpentry.org/lessons/) provides domain-specific tutorials, such as for ecology or genomics
- [Library carpentries](https://librarycarpentry.org/lessons/) contain some useful tutorials if you want to transform data frames, map data to each other and work effectively with data

Next, to the carpentries you will find a list of tutorials for more specific topics below.


### Getting started with bash

-   [A tutorial on using bash and an HPC](https://ndombrowski.github.io/cli_workshop/)
-   [Version control with git](https://github.com/fkariminejadasl/ml-notebooks/blob/main/tutorial/git.md)
-   [A tutorial on using AWK](https://ndombrowski.github.io/AWK_tutorial/), a command line tool for filtering tables, extracting patterns, etc... If you want to follow this tutorial then you can download the required input files from [here](https://github.com/ndombrowski/AWK_tutorial/tree/main/1_Inputfiles)

### Using R

-   [An R cookbook](https://ndombrowski.github.io/R_cookbook/) including some [example files](https://github.com/ndombrowski/R_cookbook/tree/main/data) if you want to code along
-   [Tutorial on data manipulation with dplyr](https://ndombrowski.github.io/Tidyverse_tutorial/)
-   [Tutorial on data visualization with ggplot2](https://ndombrowski.github.io/Ggplot_tutorial/)

### Bioinformatic workflows

-   [From sequence file to OTU table with Qiime](source/Qiime/3_evelyn_tutorial_notes.qmd)
-   [Analysing an OTU table with R](source/Qiime/OTU_table_analysis.qmd)
-   [Assembling a metagenome](https://ndombrowski.github.io/Assembly_tutorial/)
-   [Metagenomic binning](https://ndombrowski.github.io/Binning_tutorial//)
-   [Annotating microbial genomes](https://github.com/ndombrowski/Annotation_workflow)
-   [How to do generate a species tree](https://ndombrowski.github.io/Phylogeny_tutorial/)
-   [Accessing data from NCBI](source/core_tools/ncbi.qmd)

## Bioinformatic tools A-Z

-   [ATLAS](source/metagenomics/atlas.qmd): a metagenomic pipeline for QC, assembly binning and annotation
-   [Bowtie2](source/core_tools/bowtie.qmd):  A tool for aligning sequencing reads to genomes and other reference sequences
-   [CheckM2](source/metagenomics/checkm2.qmd): A tool to assess the quality of a genome assembly 
-   [Chopper](source/nanopore/chopper.qmd): A tool for quality filtering of long read data
-   [DeSeq2](source/metatranscriptomics/DSEq2.qmd): Analyse gene expression data in R
-   [Dnaapler](source/metagenomics/dnaapler.qmd): A tool to re-orient a genome, for example at dnaA
-   [FAMA](source/metagenomics/fama_readme.qmd): A fast pipeline for functional and taxonomic analysis of metagenomic sequences
-   [FastP](source/metatranscriptomics/fastp.qmd): A tool for fast all-in-one preprocessing of FastQ files
-   [FastQC](source/metagenomics/fastqc_readme.qmd): A quality control tool for read sequencing data
-   [FeatureCounts](source/core_tools/featurecounts.qmd): A read summarization program that counts mapped reads for genomic features
-   [Filtlong](source/nanopore/filtlong.qmd): A A tool for filtering long reads
-   [Flye](source/metagenomics/flye.qmd): A de novo assembler for single-molecule sequencing reads 
-   [GTDB_tk](source/metagenomics/gtdb.qmd): A software toolkit for assigning objective taxonomic classifications to bacterial and archaeal genomes
-   [HMMER](source/core_tools/hmmer.qmd): A tool for searching sequence databases for sequence homologs, and for making sequence alignments
-   [Interproscan](source/metagenomics/interproscan_readme.qmd): A tool to scan protein and nucleic sequences against InterPro signatures
-   [ITSx](source/ITSx/itsx_readme.qmd): A tool to extract ITS1 and ITS2 subregions from ITS sequences
-   [Kraken2](source/classification/kraken2.qmd): A taxonomic sequence classifier using kmers
-   [METABOLIC](source/metagenomics/metabolic.qmd): A tool to predict functional trait profiles in genome datasets
-   [MOTUS](source/metagenomics/motus.qmd): A tool to estimate microbial abundances in Illumina and Nanopore sequencing data
-   [Minimap2](source/classification/minimap2.qmd): A program to align DNA or mRNA sequences against a reference database
-   [NanoClass2](source/nanopore/nanoclass.qmd): A taxonomic meta-classifier for long-read 16S/18S rRNA gene sequencing data
-   [NanoITS](source/nanopore/nanoITS.qmd): A taxonomic meta-classifier for long-read ITS operon sequencing data
-   [Nanophase](source/nanopore/nanophase_how_to.qmd): A pipeline to generate MAGs using Nanopore long and Illumina short reads from metagenomes
-   [NanoPlot](source/nanopore/nanoplot_readme.qmd): Plotting tool for long read sequencing data
-   [NanoQC](source/nanopore/nanoqc_readme.qmd): A quality control tool for long read sequencing data
-   [Porechop](source/nanopore/porechop_readme.qmd): A tool for finding and removing adapters from Nanopore reads
-   [Prokka](source/core_tools/prokka.qmd): A tool to annotate bacterial, archaeal and viral genomes
-   [Pseudofinder](source/metagenomics/pseudofinder.qmd): A tool that detects pseudogene candidates from annotated genbank files of bacterial and archaeal genomes
-   [QUAST](source/metagenomics/quast.qmd): A Quality Assessment Tool for Genome Assemblies
-   [Samtools](source/core_tools/samtools.qmd): A tool to manipulating alignments in SAM/BAM format
-   [SeqKit](source/core_tools/seqkit.qmd): A tool for FASTA/Q file manipulation
-   [SortMerNa](source/metatranscriptomics/sortmerna.qmd): A tool to filter ribosomal RNAs in metatranscriptomic data
-   [Trinity](source/metatranscriptomics/trinity.qmd): A tool to assemble transcript sequences from Illumina RNA-Seq data   



## Bioinformatic toolbox

- [For-loops-in-bash](source/core_tools/bash-for-loops.qmd): How can I scale my research and run software not only on one but multiple files?


## Useful databases A-Z

- [KOfam](source/core_tools/kegg_db.qmd)
- [Pfam](source/core_tools/pfam_db.qmd)
- [PGAP database](source/core_tools/pgap_db.qmd)
- [TIGRFAM](source/core_tools/tigrfam_db.qmd)