---
code-block-bg: true
code-block-border-left: "#31BAE9"
execute:
  eval: false
engine: knitr
bibliography: references.bib
---

<div style="text-align: justify">

## HMMER

### Introduction

HMMER is used for searching sequence databases for sequence homologs, and for making sequence alignments [@Eddy2011]. It implements methods using probabilistic models called profile hidden Markov models (profile HMMs).

HMMER is often used together with a profile database, such as Pfam or many of the databases that participate in Interpro. But HMMER can also work with query sequences, not just profiles, just like BLAST. For example, you can search a protein query sequence against a database with phmmer, or do an iterative search with jackhmmer.

HMMER is designed to detect remote homologs as sensitively as possible, relying on the strength of its underlying probability models. In the past, this strength came at significant computational expense, but as of the new HMMER3 project, HMMER is now essentially as fast as BLAST.

For more information, please visit the [official website](http://hmmer.org/) and have a look at the [User's guide](http://eddylab.org/software/hmmer/Userguide.pdf).

### Installation

Installed on crunchomics: Yes,HMMER 3.3.2 is installed

### Usage

HMMER is a software suite with a lot of programs and possibilities, for a full list have a look at the [User's guide](http://eddylab.org/software/hmmer/Userguide.pdf). Some key programs are used to:

-   Build models and align sequences (DNA or protein)
    -   hmmbuild - Build a profile HMM from an input multiple alignment.
    -   hmmalign - Make a multiple alignment of many sequences to a common profile HMM
-   Search protein queries against protein database
    -   phmmer - Search a single protein sequence against a protein sequence database. (BLASTP-like)
    -   jackhmmer - Iteratively search a protein sequence against a protein sequence database. (PSIBLAST-like)
    -   hmmsearch - Search a protein profile HMM against a protein sequence database.
    -   hmmscan - Search a protein sequence against a protein profile HMM database.
    -   hmmpgmd - Search daemon used for hmmer.org website.

Notice, that either hmmsearch or hmmscan can compare a set of profiles to a set of sequences. Due to disk access patterns of the two tools, it is usually more efficient to use hmmsearch, unless the number of profiles greatly exceeds the number of sequences.

Examples:

#### Generate your own hmm for searches

In some cases a good hmm might not exist but you can easily create your own. For this, collect all proteins for your orthologue of interest. For example, we might want to build a custom profile for a methyl coenzyme M reductase (mcrA). To do this, search for protein sequences in databases, such as NCBI and generate a protein fasta file. Next, you can generate a multiple sequence (MSA) alignment with tools such as MAFFT, let's call this file mcrA.aln. For good alignments, consider to manually inspect it and discard sequences that don't align well.

Then you can run:

```{bash}
#generate a hmm profile
hmmbuild mcrA.hmm mcrA.aln

#we can then use this profile to against a protein database,
#my_proteins.faa could be proteins from a genome for which we want to check for the presence of mcrA proteins
hmmsearch mcrA.hmm my_proteins.faa > mcrA.out
```

#### Compare proteins of interest against a hmm database

```{bash}
hmmsearch \
	--tblout results/sequence_results.txt \
	--domtblout results/domain_results.txt \
	--notextw \
	--cpu 10 \
	KO_db.hmm \
	results/prokka/bin5.faa
```

Let's do this in an advanced mode and parse a few things afterwards. This assumes that you already have a few genomes ready and that they are found in the folder `results/faa`

```{bash}
mkdir results/faa/renamed
mkdir results/kos/

#combine proteins from different genomes in one database 
#add the filename into the fasta header and store them in our new folder
#combining all genomes into 1 file has the benefit that you only need to run hmmsearch 1x
#for this to work proberly its a good practice to add the genomeID into the fasta header so that you can easily distinguish from where each protein originally came
for i in  results/faa/*faa; do 
filename=$(basename $i .faa)
awk -v fname="$filename" '/>/{sub(">","&"fname"-");sub(/\.faa/,x)}1' $i > results/faa/renamed/$filename.faa
done

#after ensuring our files have good headers, we can combine the individual faa files 
cat results/faa/renamed/*.faa > results/faa/combined.faa

#now we can run the hmmsearch
hmmsearch --tblout results/kos/sequence_results.txt \
  --domtblout results/kos/domain_results.txt \
  --notextw --cpu 20 \
  /zfs/omics/projects/bioinformatics/databases/kegg/release_2024_26-04/KO_db.hmm \
  results/faa/combined.faa

#format the full table by:
#remove hash symbols and replacing spaces with a tab
#retain only the columns with the protein-id, ko-id, e-value, bitscpre
#and only select hits above a certain e-value
sed 's/ \+ /\t/g' results/kos/sequence_results.txt | sed '/^#/d'| sed 's/ /\t/g'| awk -F'\t' -v OFS='\t' '{print $1, $3, $6, $5}' | awk -F'\t' -v OFS='\t' '($4 + 0) <= 1E-3'  > results/kos/sequence_results_red_e_cutoff.txt

#get best hit/protein based on bit score, and e-value
#i.e. one protein might hit toward more than 1 HMM, we only want to retain one hit
sort -t$'\t' -k3,3gr -k4,4g results/kos/sequence_results_red_e_cutoff.txt | sort -t$'\t' --stable -u -k1,1  | sort -t$'\t' -k3,3gr -k4,4g >  results/kos/sequence_results_red_e_cutoff_best_hit.txt

#merge with KO mapping file 
#ensure that the KOs are in column 2 and 1 of the results and the mapping file
#-o decides what columns are kept after the merge from table 1 and 2. I.e. from the mapping file only columns 2 and 12 are used
LC_ALL=C join -a1 -1 2 -2 1 -e'-' -t $'\t' -o1.1,1.2,1.4,1.3,2.2,2.12 <(LC_ALL=C sort -k2 results/kos/sequence_results_red_e_cutoff_best_hit.txt) <(LC_ALL=C sort -k1  /zfs/omics/projects/bioinformatics/databases/kegg/release_2024_26-04/ko_list) | LC_ALL=C  sort >  results/kos/temp1

#add in an extra column that lists whether hits have a high confidence score
awk  -v OFS='\t' '{ if ($4 > $5){ $7="high_score" }else{ $7="-" } print } ' results/kos/temp1 > results/kos/temp2

#add header
echo -e "accession\tKO_hmm\tKO_e_value\tKO_bit_score\tKO_bit_score_cutoff\tKO_Definition\tKO_confidence" | cat - results/kos/temp2 > results/kos/KO_hmm.tsv
```


Output:

-   The `--tblout` output option produces the target hits table. The target hits table consists of one line for each different query/target comparison that met the reporting thresholds, ranked by decreasing statistical significance (increasing E-value). Page 67 of the [User's guide](http://eddylab.org/software/hmmer/Userguide.pdf) explains each column in more detail.
-   In protein search programs, the --domtblout option produces the domain hits table. There is one line for each domain. There may be more than one domain per sequence. Page 70 of the [User's guide](http://eddylab.org/software/hmmer/Userguide.pdf) explains each column in more detail.
-   `--notextw`: Unlimits the length of each line in the main output. The default is a limit of 120 characters per line, which helps in displaying the output cleanly on terminals and in editors, but can truncate target profile description lines.

For more options, check the manual or use `hmmsearch -h`