---
code-block-bg: true
code-block-border-left: "#31BAE9"
execute:
  eval: false
engine: knitr
bibliography: references.bib
---

<div style="text-align: justify">

## PGAP

### Introduction

The PGAP database is a database originally based on the TIGRFAMs database [@li2021].

The original TIGRFAMs database was a research project of The Institute for Genomic Research (TIGR) and its successor, the J. Craig Venter Institute ([JCVI](https://www.jcvi.org/)) [@haft2003]. TIGRFAMs is a collection of manually curated protein families focusing primarily on prokaryotic sequences. It consists of hidden Markov models (HMMs), multiple sequence alignments, Gene Ontology (GO) terminology, Enzyme Commission (EC) numbers, gene symbols, protein family names, descriptive text, cross-references to related models in TIGRFAMs and other databases, and pointers to literature.

The TIGRFAMs database was transferred in April 2018 to the National Center for Biotechnology Information (NCBI), which now holds the creative commons license to this data and is responsible for maintaining and distributing this intellectual property. The database is used in NCBI's Prokaryotic Genome Annotation Pipeline for GenBank and RefSeq sequence annotation, and curators continue to revise existing models.

Notice: Release 15.0 (January 2013) was the last full release of TIGRFAMs from JCVI and newer versions are maintained by NCBI. The up-to-date versions of all TIGRFAM models are available for download by [FTP](https://ftp.ncbi.nlm.nih.gov/hmm/current/) as a component of the [current release of PGAP HMMs](https://ftp.ncbi.nlm.nih.gov/hmm/current/). They are recognizable by an accession number beginning with “TIGR”, or by the source designation “JCVI”

### Installation

Available on crunchomics: Yes,

-   PGAP is installed as part of the bioinformatics share. If you have access to crunchomics and have not yet access to the bioinformatics you can send an email with your Uva netID to [Nina Dombrowski](n.dombrowski@uva.nl).

The PGAP database can be found here:

-   `/zfs/omics/projects/bioinformatics/databases/pgap/release_15.0`.

If you want to download the database yourself, you can do:

```{bash}
wget https://ftp.ncbi.nlm.nih.gov/hmm/current/hmm_PGAP.tsv
wget https://ftp.ncbi.nlm.nih.gov/hmm/current/hmm_PGAP.LIB

#cleanup mapping file 
sed -i "s/ /_/g" hmm_PGAP.tsv
```