---
code-block-bg: true
code-block-border-left: "#31BAE9"
execute:
  eval: false
engine: knitr
bibliography: references.bib
---

<div style="text-align: justify">

## GTDB_tk

### Introduction

GTDB_Tk is a software toolkit for assigning objective taxonomic classifications to bacterial and archaeal genomes [@chaumeil2019]. It uses the GTDB database to assign your genome(s) of interest to a taxonomy [@parks2020]. For more information, visit the tool\`s [github](https://github.com/Ecogenomics/GTDBTk) and [website](https://ecogenomics.github.io/GTDBTk/).

### Installation

Installed on crunchomics: Yes,

-   GTDB_Tk v2.3.2 is installed as part of the bioinformatics share. If you have access to crunchomics and have not yet access to the bioinformatics you can send an email with your Uva netID to [Nina Dombrowski](n.dombrowski@uva.nl).
-   Additionally, the GTDB r214 database that is compatible with GTDB_Tk v2.3.2 is installed on the metatools share. To get access to metatools, send an email with your Uva netID to Anna Heintz Buschart [a.u.s.heintzbuschart\@uva.nl](mailto:a.u.s.heintzbuschart@uva.nl){.email}.

After you were added to the bioinformatics and metatools share, you can add the conda environments that are installed in these shares as follows (if you have already done this in the past, you don't need to run these commands):

```{bash}
conda config --add envs_dirs /zfs/omics/projects/bioinformatics/software/miniconda3/envs/
conda config --add envs_dirs /zfs/omics/projects/metatools/TOOLS/miniconda3/envs/
```

If you want to install it yourself, you can run:

GTDB_tk 2.3.2:

```{bash}
mamba create --name gtdbtk_2.3.2 -c bioconda gtdbtk=2.3.2

#download the GDTB database (beware its quite large)
cd my_download_folder
wget https://data.gtdb.ecogenomic.org/releases/release214/214.0/auxillary_files/gtdbtk_r214_data.tar.gz 
tar xvzf gtdbtk_r214_data.tar.gz

#link data to gtdbtk
conda activate gtdbtk_2.3.2
conda env config vars set GTDBTK_DATA_PATH="my_download_folder/GTDB_tk_r214";
conda deactivate

#check that the install went ok
conda activate gtdbtk_2.3.2
gtdbtk check_install
conda deactivate
```

GTDB_tk 2.4.0:

```{bash}
mamba create -p /zfs/omics/projects/bioinformatics/software/miniconda3/envs/gtdbtk_2.4.0 -c bioconda gtdbtk=2.4.0

#get gtdb data, v220 
cd /zfs/omics/projects/bioinformatics/databases/gtdb
wget https://data.gtdb.ecogenomic.org/releases/latest/auxillary_files/gtdbtk_package/full_package/gtdbtk_data.tar.gz
tar xvzf gtdbtk_data.tar.gz 

#link data to gtdbtk
conda activate gtdbtk_2.4.0
conda env config vars set GTDBTK_DATA_PATH="/zfs/omics/projects/metatools/DB/GTDB_tk/GTDB_tk_r214";
conda deactivate
```

### Usage

```{bash}
mkdir results/gtdb 

conda activate gtdbtk_2.3.2

gtdbtk classify_wf --genome_dir  data/genomes/fna/ \
  --out_dir results/gtdb \
   --mash_db /zfs/omics/projects/metatools/DB/GTDB_tk/GTDB_tk_r214/mash/ \
  --cpus 20

conda deactivate
```

The file `gtdbtk.bac120.summary.tsv` will contain key information about the taxonomic assignment of your genome(s).

For a full set of options and description how the tool works, please visit the [manual](https://ecogenomics.github.io/GTDBTk/).


### Common Issues and Solutions

-   **Issue 1:** Running out of memory as described [here](https://ecogenomics.github.io/GTDBTk/faq.html)
    -   Solution 1: Use `--scratch_dir` and `--pplacer_cpus 1`
-   **Issue 2:** Using a gtdbtk version with the incorrect database
    -   Solution 2: Ensure that you use the right gtdbtk-db combination, as listed [here](https://ecogenomics.github.io/GTDBTk/installing/index.html)