---
code-block-bg: true
code-block-border-left: "#31BAE9"
execute:
  eval: false
engine: knitr
bibliography: references.bib
---

<div style="text-align: justify">

## pycirclize

### Introduction

pyCirclize is a circular visualization python package based on matplotlib [@shimoyama2022]. This package is developed for the purpose of easily and beautifully plotting circular figure such as Circos Plots and Chord Diagrams in Python. In addition, useful genome and phylogenetic tree visualization methods for the bioinformatics field are also implemented.

### Installation

Installed on Crunchomics: Yes,

-   The pycirclize conda environment is part of the bioinformatics share. If you have access to Crunchomics and have not yet access to the bioinformatics share, then you can send an email with your Uva netID to Nina Dombrowski, n.dombrowski\@uva.nl.
-   Afterwards, you can add the bioinformatics share as follows (if you have already done this in the past, you don't need to run this command):

```{bash}
conda config --add envs_dirs /zfs/omics/projects/bioinformatics/software/miniconda3/envs/
```

If you want to install it yourself, you can run:

```{bash}
mamba create -p pycirclize -c bioconda -c conda-forge python=3.10.12 pycirclize numpy matplotlib
```

### Usage

You can easily generate custom plots using pycirclize in python after activating the environment with `conda activate pycirclize`. To get started, you can view more instructions [here](https://moshi4.github.io/pyCirclize/).

Additionally, we provide a small python wrapper, that can take a Prokka gbk (possibly others, however, that was not tested yet) and generate a circos plot that will look something like this:

![](/img/circos_plot.png){width="500"}

```{bash}
conda activate pycirclize

python /zfs/omics/projects/bioinformatics/scripts/generate_circos_plot.py \
	-i dnaapler.gbk \
	-o circos_plot.pdf \
	-g genes_of_interest.txt

conda deactivate
```

Settings:

-   `-h`, `--help` show this help message and exit
-   `-i` INPUT, `--input` INPUT Input GenBank file generated by Prokka (.gbk)
-   `-o` OUTPUT, `--output` OUTPUT Output PDF file (.pdf)
-   `-g` GENES, `--genes` GENES Text file containing prokka IDs of genes of interest that should be added to the plot (one gene per line). This setting is optional.
-   `--version` show program's version number and exit